ethdb/privateapi/ethbackend.go

```diff
@@ -85,10 +90,12 @@ type PayloadMessage struct {
 
 func NewEthBackendServer(ctx context.Context, eth EthBackend, db kv.RwDB, events *Events, blockReader interfaces.BlockReader,
 	config *params.ChainConfig, reverseDownloadCh chan<- PayloadMessage, statusCh <-chan ExecutionStatus, waitingForBeaconChain *uint32,
+	skipCycleHack chan struct{}, assemblePayloadPOS assemblePayloadPOSFunc, proposing bool,
 ) *EthBackendServer {
 	return &EthBackendServer{ctx: ctx, eth: eth, events: events, db: db, blockReader: blockReader, config: config,
 		reverseDownloadCh: reverseDownloadCh, statusCh: statusCh, waitingForBeaconChain: waitingForBeaconChain,
-		pendingPayloads: make(map[uint64]types2.ExecutionPayload),
+		pendingPayloads: make(map[uint64]types2.ExecutionPayload), skipCycleHack: skipCycleHack,
+		assemblePayloadPOS: assemblePayloadPOS, proposing: proposing, condPauseAssemble: *sync.NewCond(&sync.Mutex{}),
```

Comment:
You don't need a separate mutex for condPauseAssemble: just re-use `mu`.

Reply:
We cannot do that as reusing mu, would cause mu.Lock be called multiple times without any unlocks aka. stalling everything. the reason of this is that what `Wait` essentially does is locking the mutex, so we cannot reuse the same mutex. same goes with `Broadcast`, which essentially unlocks the mutex.

Reply:
I've double checked https://pkg.go.dev/sync#Cond.Wait and the advise is to check for a condition in a loop. So we can use a single `sync.Cond` (and its mutex) for everything, just with different checks. I've submitted https://github.com/ledgerwatch/erigon/pull/3187/commits/4de1c30397706b69319dc75e4623d65f8eecbd3e to that effect.